### PR TITLE
Gui part 2

### PR DESCRIPTION
**Pull request by thinkslynk**

### CODE REVIEW

File: src/main/kotlin/com/refinedmods/refinedstorage/RSGui.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import com.refinedmods.refinedstorage.block.ConstructorBlock
+import com.refinedmods.refinedstorage.gui.screenhandlers.AlternativesScreenHandler
+import com.refinedmods.refinedstorage.gui.screenhandlers.AmountScreenHandler
+import com.refinedmods.refinedstorage.gui.screenhandlers.ConstructorScreenHandler
+import com.refinedmods.refinedstorage.gui.screenhandlers.FilterScreenHandler
+import net.fabricmc.fabric.api.screenhandler.v1.ScreenHandlerRegistry
+import net.minecraft.item.ItemStack
+import net.minecraft.util.Identifier
+
+object RSGui {
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/RSGui.kt:12]: Sounds like we need an annotation for this =P

File: src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/network/NetworkNodeManager.kt [hunk 2]
```diff
@@ -5,81 +5,89 @@ import com.refinedmods.refinedstorage.api.network.node.INetworkNodeFactory
...
-//        }
-//
 //        // Lookup and cache
 //        return BlockComponents.get(RSComponents.NETWORK_NODE_PROXY, world, pos)?.let {
 //            nodes[pos] = it.node
 //            it.node
 //        }
 
+        // Use cache if it exists
+        if(nodes.contains(pos)) {
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/network/NetworkNodeManager.kt]: Shouldn't we need to verify that this is still a valid node?

File: src/main/kotlin/com/refinedmods/refinedstorage/block/CableBlock.kt [hunk 3]
```diff
@@ -25,6 +25,7 @@ import net.minecraft.world.BlockView
 import net.minecraft.world.World
 import net.minecraft.world.WorldAccess
 import java.util.function.Function
+import net.minecraft.entity.player.PlayerEntity
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/block/CableBlock.kt]: How come this one needs to be imported now?

File: src/main/kotlin/com/refinedmods/refinedstorage/container/GridContainer.kt [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+//    override fun canMergeSlot(stack: ItemStack?, slot: Slot): Boolean {
+//        return if (slot === craftingResultSlot || slot === patternResultSlot) {
+//            false
+//        } else super.canMergeSlot(stack, slot)
+//    }
+//
+//    init {
+//        grid!!.addCraftingListener(this)
+//    }
+//}
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/container/GridContainer.kt]: How to make your prs look bigger than they are 101: add file that is completely commented out

=P

File: src/main/kotlin/com/refinedmods/refinedstorage/container/slot/BaseSlot.kt [hunk 5]
```diff
@@ -13,6 +13,8 @@ open class BaseSlot(itemHandler: Inventory, inventoryIndex: Int, x: Int, y: Int)
         return this
     }
 
-    open val isEnabled: Boolean
-        get() = enableHandler.get()
+    open val isEnabled: Boolean = enableHandler.get()
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/container/slot/BaseSlot.kt]: So this will never change?

File: src/main/kotlin/com/refinedmods/refinedstorage/container/slot/filter/FilterSlot.kt [hunk 6]
```diff
@@ -1,20 +1,35 @@
...
+        val isBlockItem = stack.item is BlockItem
 
-        return if (super.canInsert(stack)) {
-            if (isBlockAllowed) {
-                stack.item is BlockItem
-            } else true
+        return if (superCan) {
+            if (isBlockAllowed) { isBlockItem } else true
         } else false
     }
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/container/slot/filter/FilterSlot.kt]: Why did you change this?

File: src/main/kotlin/com/refinedmods/refinedstorage/data/BaseBlockEntityData.kt [hunk 7]
```diff
@@ -1,42 +1,56 @@
...
         }
 
     companion object{
+        val ID = Identifier(RS.ID, "base_block_entiy_data")
         // TODO Add annotation to Fabric Seams for TrackedDataHandlers
         val SERIALIZER = object: TrackedDataHandler<BaseBlockEntityData>{
             override fun write(buf: PacketByteBuf, o: BaseBlockEntityData) {
                 buf.writeBlockPos(o.blockPos)
-                buf.writeEnumConstant(o.direction)
+                buf.writeBoolean(o.direction != null)
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/data/BaseBlockEntityData.kt:37]: I made an extension function for `optionalUuid` might also use that here?

File: src/main/kotlin/com/refinedmods/refinedstorage/data/sync/BiSyncedData.kt [hunk 8]
```diff
@@ -59,14 +64,14 @@ open class BiSyncedData<T>(
 
     override fun send() {
         val byteBuffer = byteBuffers.buffer()
-        try {
+//        try {
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/data/sync/BiSyncedData.kt]: errr, should this be here?

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt [hunk 9]
```diff
@@ -1,45 +1,54 @@
...
     protected abstract val defaultAmount: Int
     protected abstract fun canAmountGoNegative(): Boolean
     protected abstract val maxAmount: Int
-    protected open val amountPos: Pair<Int, Int>
-        protected get() = Pair.of(7 + 2, 50 + 1)
-    protected open val okCancelPos: Pair<Int?, Int?>?
-        protected get() = Pair.of(114, 33)
-    protected open val okCancelButtonWidth: Int
-        protected get() = 50
+    protected open val amountPos: Pair<Int, Int> = Pair.of(7 + 2, 50 + 1)
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt]: Why are we using this `Pair` class?

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt [hunk 10]
```diff
@@ -59,23 +68,29 @@ abstract class AmountSpecifyingScreen<T : ScreenHandler>(
         val increments = increments
         var xx = 7
         val width = 30
-        for (i in 0..2) {
+        (0..2).forEach { i ->
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt]: `repeat(3)`

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt [hunk 11]
```diff
@@ -85,53 +100,36 @@ abstract class AmountSpecifyingScreen<T : ScreenHandler>(
...
-        if (newAmount > maxAmount) {
-            newAmount = maxAmount
-        }
-        amountField!!.text = newAmount.toString()
-    }
+        val oldAmount = amountField.text.toIntOrNull() ?: 0
+        val newAmount = if (canAmountGoNegative())
+            oldAmount + increment
+        else
+            1.coerceAtLeast((if (oldAmount == 1 && increment != 1) 0 else oldAmount) + increment)
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt]: This is confusing, normally you pass the constant as the parameter and call it on the variable amount.

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+
+abstract class BaseScreen<T : ScreenHandler>(
+    container: T,
+    player: PlayerEntity,
+    title: Text
+): HandledScreen<T>(container, player.inventory, title) {
+    private var sideButtonY = 0
+
+    fun getX(): Int = x
+    fun getY(): Int = y
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt:46]: Why do these exist?

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screenhandlers/ConstructorScreenHandler.kt [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+    // TODO OLD
+    init {
+        syncedData.register()
+
+        buf?.let {
+            syncedData.load(it)
+        }
+
+        (0 until 4).forEach {
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/gui/screenhandlers/ConstructorScreenHandler.kt]: `repeat(5)`

**Review by Kroppeb — COMMENTED** [review]
Initial review, Gonna have a more in-depth later

File: src/main/kotlin/com/refinedmods/refinedstorage/RSGui.kt (hunk 1 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/RSGui.kt:12]: I agree, I had some `todo`s for this. But I was waiting to add any new annotations until your stuff is merged.

File: src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/network/NetworkNodeManager.kt (hunk 2 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/network/NetworkNodeManager.kt]: Maybe. It might be best to move the `markedForRemoval` flag into the node and check that. I was avoiding re-looking up the entity via the world (which would defeat the point of the cache.)

File: src/main/kotlin/com/refinedmods/refinedstorage/container/GridContainer.kt (hunk 4 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/container/GridContainer.kt]: Gotta make it look important or no one will review it!  ;P

File: src/main/kotlin/com/refinedmods/refinedstorage/block/CableBlock.kt (hunk 3 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/block/CableBlock.kt]: It doesn't. Not sure how that happened. I'll remove it, thanks.

File: src/main/kotlin/com/refinedmods/refinedstorage/container/slot/BaseSlot.kt (hunk 5 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/container/slot/BaseSlot.kt]: `enableHandler` is a var, so it it's up to the setter.

File: src/main/kotlin/com/refinedmods/refinedstorage/container/slot/filter/FilterSlot.kt (hunk 6 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/container/slot/filter/FilterSlot.kt]: While I was debugging some issues related to the slots, I moved the boolean logic into vars to log them. I removed the log but didn't switch it back. I can if you prefer.

File: src/main/kotlin/com/refinedmods/refinedstorage/data/BaseBlockEntityData.kt (hunk 7 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/data/BaseBlockEntityData.kt:37]: I'll take a look, thanks.

File: src/main/kotlin/com/refinedmods/refinedstorage/data/sync/BiSyncedData.kt (hunk 8 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/data/sync/BiSyncedData.kt]: Nope, I cleaned it out of the other two classes, but not this one. Will remove, thanks.

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt (hunk 10 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt]: Ooo, I like. Will change.

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt (hunk 9 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt]: Because the original code did. I only update this screen the bare minimum to get `BaseScreen` to compile. It's not currently used. Could use some TLC at some point. Is there something better than a `Pair` in your opinion?

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt (hunk 10 above)
Comment (Kroppeb) [reply src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt:71]: Kotlin pair =p

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt (hunk 9 above)
Comment (Kroppeb) [reply src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt]: Kotlin `Pair`?

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt (hunk 12 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt:46]: Because `x` and `y` are protected, but need to be read by various widget classes.

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt (hunk 9 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt]: Oh, duh. My bad.

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt (hunk 11 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt]: Yeah, it was an intelliJ recommendation; but I agree with you. I'm going to switch it back to the more readable form.

Comment (Kroppeb) [reply src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/AmountSpecifyingScreen.kt]: `coerceAtLeast` can be more readable, but in this case, switching the arguments would add more brackets

File: src/main/kotlin/com/refinedmods/refinedstorage/container/slot/BaseSlot.kt (hunk 5 above)
Comment (Kroppeb) [reply src/main/kotlin/com/refinedmods/refinedstorage/container/slot/BaseSlot.kt]: But it's no longer a getter, so it will only be checked once?

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+        val rY = mouseY - this.y
+
+        RenderSystem.color4f(1.0f, 1.0f, 1.0f, 1.0f)
+
+        renderForeground(matrixStack, rX, rY)
+        buttons.filter { it is SideButton && it.isHovered }
+            .forEach { it.renderToolTip(matrixStack, rX, rY) }
+
+        screenHandler.slots.filter { it is FluidFilterSlot && it.isEnabled }
+            .map { it as FluidFilterSlot }
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt]:
```suggestion
let{it as List<FluidFilterSlot>}
```

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/ConstructorScreen.kt [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+import net.minecraft.text.Text
+
+class ConstructorScreen(
+        gui: ConstructorScreenHandler,
+        player: PlayerEntity,
+        title: Text
+) : BaseScreen<ConstructorScreenHandler>(gui, player, title) {
+
+        override fun onPostInit(x: Int, y: Int) {
+                this.x -= 60
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/ConstructorScreen.kt]: Could you explain what this does?

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+        y: Int,
+        w: Int,
+        h: Int,
+        text: Text?,
+        enabled: Boolean,
+        visible: Boolean,
+        onPress: ButtonWidget.PressAction?
+    ): ButtonWidget  {
+        val button = ButtonWidget(x, y, w, h, text, onPress)
+        button.active
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt]: what?

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+        button.y = this.y + sideButtonY
+        sideButtonY += SideButton.HEIGHT + 2
+        addButton(button)
+    }
+
+    fun bindTexture(namespace: String, filenameInTexturesFolder: String) {
+        TrackedDataHandlerRegistry.FLOAT
+        MinecraftClient.getInstance().textureManager
+            .bindTexture(
+                TEXTURE_CACHE.computeIfAbsent("$namespace:$filenameInTexturesFolder")
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt:301]: What is this for?

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/ConstructorScreen.kt (hunk 15 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/ConstructorScreen.kt]: Test code that should be removed (Was trying to shift the gui over so that REI didn't overlap, but it doesn't help.)

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt (hunk 17 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt:301]: *shurg* 
https://github.com/refinedmods/refinedstorage/blob/aba679a6b0f56f65bb9f14272815377db225df18/src/main/java/com/refinedmods/refinedstorage/screen/BaseScreen.java#L276``

File: src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt (hunk 16 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/gui/screen/BaseScreen.kt]: (Removing)

File: src/main/kotlin/com/refinedmods/refinedstorage/container/slot/BaseSlot.kt (hunk 5 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/container/slot/BaseSlot.kt]: I don't know why I thought that would delegate to the function. My bad. Will switch back.

File: src/main/kotlin/com/refinedmods/refinedstorage/data/BaseBlockEntityData.kt (hunk 7 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/data/BaseBlockEntityData.kt:37]: I tried to find that extension, but couldn't. Is it on anther branch? Or am I just blind? lol

File: src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/network/NetworkNodeManager.kt (hunk 2 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/network/NetworkNodeManager.kt]: This has been updated, but could do with another review.

File: src/main/kotlin/com/refinedmods/refinedstorage/data/BaseBlockEntityData.kt (hunk 7 above)
Comment (Kroppeb) [reply src/main/kotlin/com/refinedmods/refinedstorage/data/BaseBlockEntityData.kt:37]: I found them in `PacketBufUtils`
